internal/webhook/release_webhook.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2024
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package webhook
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	hmcv1alpha1 "github.com/Mirantis/hmc/api/v1alpha1"
+)
+
+type ReleaseValidator struct {
+	client.Client
+}
+
+// SetupWebhookWithManager will setup the manager to manage the webhooks
+func (v *ReleaseValidator) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	v.Client = mgr.GetClient()
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(&hmcv1alpha1.Release{}).
+		WithValidator(v).
+		Complete()
+}
+
+var _ webhook.CustomValidator = &ReleaseValidator{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (*ReleaseValidator) ValidateCreate(_ context.Context, _ runtime.Object) (admission.Warnings, error) {
+	return nil, nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (*ReleaseValidator) ValidateUpdate(_ context.Context, _, _ runtime.Object) (admission.Warnings, error) {
+	return nil, nil
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (v *ReleaseValidator) ValidateDelete(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	release, ok := obj.(*hmcv1alpha1.Release)
+	if !ok {
+		return admission.Warnings{"Wrong object"}, apierrors.NewBadRequest(fmt.Sprintf("expected Release but got a %T", obj))
+	}
+	mgmtList := &hmcv1alpha1.ManagementList{}
+	if err := v.List(ctx, mgmtList); err != nil {
+		return nil, err
+	}
+	if len(mgmtList.Items) != 1 {
+		return nil, fmt.Errorf("expected 1 Management object, got %d", len(mgmtList.Items))
+	}
```

Comment:
Maybe it's worth allowing the Release deletion if there is no `Management` object yet? Ex. the HMC was installed but the Management object was not yet created, the new `Release` was published and the user wants to create a new release and remove an old one and then, create the `Management` object. What do you think?

Reply:
It wouldn't harm I guess, but realistically admission is not enabled until the `Management` object is reconciled, so should not really happen